### ISSUE TITLE
Fix permissions of binary produced by Docker

### ISSUE THREAD

**Issue by mariaschett** [project-oak/transparent-release#129]
Context: following up on the discussion in #123 .

Background: The binary produced by Docker is owned by `root`. I believe this causes the `builder` to fail [here](https://github.com/project-oak/transparent-release/blob/d98c4b9b3b02e4a4c8ec24b6ae1c4d001001da14/internal/builder/builder.go#L53) with `Couldn't build the binary: couldn't build the binary: missing expected output file in [...]`

As one solution: I propose to follow the solution in Oak. On the example of the hello-transparent-release repo this would require  #123  together with [changing the owner](https://github.com/project-oak/hello-transparent-release/blob/774212d35ab19bc9e3bddec1ac2b27878158f2e3/scripts/build_from_docker#L18).

However, @tiziano88 suggests a different solution: https://github.com/project-oak/transparent-release/pull/123#discussion_r960548614 .

I think, the first question is: do we want to run the `builder` with `sudo`?

@rbehjati @tiziano88

**Comment by rbehjati** [comment]
What command do you use for building the binary? Are you sure the file gets built? I am asking because I experimented with the builder a bit, and when there is a permission issue, the error message contains "permission denied".

**Comment by tiziano88** [comment]
I think we need to separate two different use cases, both of which currently rely on Docker (and in fact the same image):

- local development (e.g. running VScode with a Docker dev container): for this use case, it is important that the resulting artifacts have the same UID/GID as the real user, because the workflow usually relies on running some commands from the docker image, while a volume from the external hard drive is mounted inside it, so that the output binaries can then be used on the host machine (e.g. running a server for testing); this requires at the very least a way of passing in the UID/GID information while running the command in question, but then the issue is that in principle the build may be non-deterministic based on the UID/GID (e.g. it could do something completely different if the UID is 1234, or it may accidentally embed the UID in an artifact, etc.)
- release process (including provenance / reproducibility generation): for this use case, we don't care who owns the resulting binary; note this is fine because in order to identify binaries we only hash their content, not their metadata; in particular, the name, time, permissions, and owner of a file do not contribute to its sha256 hash; therefore, in this case, I think always running the build process as a predefined user (`root` for the sake of simplicity, but it may be any other user, as long as its name and UID are fixed somewhere; for instance it may be a `docker` user with UID 1000, and maybe that's actually better than `root` -- this probably requires more discussion / experiments) may reduce sources of non-reproducibility (at the expense of not being able to easily cross the docker / host boundary for artifacts because of mismatching ownership)

To summarize, my suggestion is to try (if possible) to make the build process that is part of transparent release as hermetic as possible, but at the same time leave some way for interactive development to keep working as it does currently, in an ergonomic way.

Note this may also mean that the transparent release process shows that running a specific command should generate an expected binary hash, but then when running locally (with a different UID), a different binary is generated; this will happen in some cases, and it needs to be addressed, but the important point is that the transparent release process should be the authority in that case (since any other UID may generated a completely different binary, at least in principle).

**Comment by mariaschett** [comment]
I am running 
```go run cmd/builder/main.go -build_config_path /home/mschett/src/hello-transparent-release/buildconfigs/hello_transparent_release.toml
```

The command in the build_config is actually `command = `["bazel", "build", "//:HelloTransparentRelease"]` (as e.g., [here](https://github.com/project-oak/hello-transparent-release/blob/85aae7ab14eae6f18a8dc81afd85b76ade0558b1/buildconfigs/hello_transparent_release.toml#L14) in the earlier commit. 

The error is `Couldn't build the binary: couldn't build the binary: missing expected output file in ./out/HelloTransparentRelease`. 

I can find the binary in the `tmp/bazel-bin` folder (owned by root), but IIUC the copying to the `out` folder fails.

**Comment by mariaschett** [comment]
> * we don't care who owns the resulting binary

Yes, but the IIUC the `builder` _does_ care, because it can't copy the binary, if not run under root.

**Comment by tiziano88** [comment]
But I don't think the builder necessarily needs to interact with the mounted path directly (like humans do during local development) -- could it use `docker cp` or some other mechanism to get that file, regardless of who owns it inside the container?

**Comment by tiziano88** [comment]
At a closer look, it seems that [`docker cp` may not work on user mount points](https://docs.docker.com/engine/reference/commandline/cp/#corner-cases). Maybe there are other options that do not require mount points (e.g. just adding / copying the source files to the image itself), but that may require larger changes.

**Comment by mariaschett** [comment]
So I also could not immediately see how to make `docker cp` work to copy the binary from the Docker container.

Which brings me back to how the `builder` can copy the binary if not run under root. 

Question: Is it possible for the `builder` to _not_ copy the binary?

A question I think you are raising, @tiziano88: Should we fix the user in the Docker image?

> `docker` user with UID 1000, and maybe that's actually better than `root` -- this probably requires more discussion / experiments

What discussion / experiments did you have in mind?

**Comment by tiziano88** [comment]
I haven't tried it, but the suggestions on the Docker guide may still work though (though it's a bit clunky, involving `tar` and piping the content of the file over stdio).

```bash
docker exec CONTAINER tar Ccf $(dirname SRC_PATH) - $(basename SRC_PATH) | tar Cxf DEST_PATH -
```

Maybe even a simplified version without `tar` would work since we only ever want to copy a single file and not a dir.

This step may be run either at the end of the build process (following the build command), or as a separate step, but I suspect the latter may require getting hold of the container id somehow.

> A question I think you are raising, @tiziano88: Should we fix the user in the Docker image (for reducing sources of non-reproducibility)?

Yes I definitely think we need to fix the UID to a known value in any case, we cannot keep passing the UID from the host. Whether 0 or 1000 or another value is to be used, should be something that we can experiment with, although in principle it should not matter. Maybe sticking to 0/root is the simplest anyways.

> What discussion / experiments did you have in mind?

Mostly just try and run the build process for Oak or some other repo, with different UID, and see if it works first, and if it makes any difference in terms of the generated output file(s). If everything always works and produces the same exact binary all the time, then it's easy, and we can just stick to 0/root.

BTW I'm also looking into https://github.com/project-oak/oak/issues/2602 at the moment, and that may influence this discussion. Actually the nice thing about that is that the root user from inside the container would be mapped to the host user, so we would get the best of both worlds, but we don't need to worry about that for now I think.

**Comment by rbehjati** [comment]
> ```shell
> docker exec CONTAINER tar Ccf $(dirname SRC_PATH) - $(basename SRC_PATH) | tar Cxf DEST_PATH -
> ```

I am a bit lost in the discussion and in particular I am not sure who will use the command above. @tiziano88 Are you suggesting this should be added to the builder? What problem does it solve?

To clarify: The builder does not copy files, but it needs **read access** to the generated output binary that it expects to find in the given `output_path`. The read access is required because the builder needs to compute the SHA256 hash of the generated binary. 

I am not an expert in Linux + Docker permissions, but it seems that inside [`the builder`](https://github.com/project-oak/transparent-release/blob/d98c4b9b3b02e4a4c8ec24b6ae1c4d001001da14/cmd/builder/main.go) the `docker run` command is run with a user different from the user that owns and runs `the builder` process. As a result, after the build is completed, the builder process may not have the permission to read the generated binary. On GitHub everything is run as `root` so there is no problem, but when **testing** locally, we run into permission issues since we cannot run locally as root. In my experiments `bazel test` also has all the permissions it needs, but not `bazel run`!!

**Comment by rbehjati** [comment]
@mariaschett  I just noticed that your latest repo has `scripts/build_from_docker`, and it is the command called in `buildconfigs/hello_transparent_release.toml`. Do you get the same "missing expected output file in ..." error even with this? (without using the fix in #123)

**Comment by tiziano88** [comment]
You are right that the main goal is to measure (hash) the output binary, and therefore strictly speaking there is no need to copy it anywhere, but I would not want us to end up in the situation in which we run `sha256` from within the container, assuming it's installed there (since in general it won't be installed, or if it is, it may lie about the computed hash) -- I don't think this what being suggested anyways, I'm just mentioning as a side comment.

In any case perhaps copying it somewhere else first may make things simpler (I assumed this is what was happening already, but perhaps not). Once it's pulled out of the container, the host can then run whatever (trusted) tools it needs to, and it can also publish it to ent or to a GitHub release (another reason why it's useful to have the file around on the host).

Now, if we rely on the fact that the repo dir is directly mounted into the docker container, and also that the output file ends up in a path within that mount point, it may be possible for the host to just access the mounted dir from the outside, and it will find the file there. This though seems hard to achieve without mapping the UID (though it may be possible somehow). https://github.com/project-oak/oak/issues/2602 may end up helping with that because (if it works out) it would transparently map a fixed UID within the container (e.g. 0) to the UID of the user on the host, but without having to pass the UID in, which may be used to alter reproducibility.

Anyways note that in general the output path may not be within the mounted dir at all (e.g. the artifact may end up in `/something_else/out`, which is not mounted, and only exists within the container), so we need a way of pulling out an arbitrary file from the container, after the build step has finished running.

AFAICT the options on the table are either something with `docker cp`, or `docker exec` and pipe that over stdio in order to reconstruct the file on the host.

Or are working under a stricter set of assumptions, e.g. that everything happens within the mounted path?

**Comment by mariaschett** [comment]
> Do you get the same "missing expected output file in ..." error even with this? (without using the fix in #123)

Actually, yes. But maybe I am doing something wrong. Does that not make sense to you?

**Comment by rbehjati** [comment]
> Now, if we rely on the fact that the repo dir is directly mounted into the docker container, 

That is what we do.

> and also that the output file ends up in a path within that mount point, 

This is what we expect. The provider of the builder image must make sure this condition holds. 

> it may be possible for the host to just access the mounted dir from the outside, and it will find the file there. This though seems hard to achieve without mapping the UID (though it may be possible somehow).

I think finding is different from reading the file. Right? And mapping the UID is needed for read access. Is that right? 

> Anyways note that in general the output path may not be within the mounted dir at all (e.g. the artifact may end up in `/something_else/out`, which is not mounted, and only exists within the container), so we need a way of pulling out an arbitrary file from the container, after the build step has finished running.

What do you mean by "we"? Do you mean the `builder tool` or the `user` who also provides the builder docker image? My understanding is that the latter is responsible for copying the generated binary into a path relative to the root of the repo (also the mount point). This path should be specified in the `output_path`. I cannot think of any other solution that universally guarantees that the file will be accessible outside the container.

**Comment by mariaschett** [comment]
Following the PR to [use rootless docker](https://github.com/project-oak/oak/pull/3224), I'll go ahead and try whether rootless docker will work with using the `builder tool` on [hello-transparent-release](https://github.com/project-oak/hello-transparent-release)---without explicitly changing the ownership. 

As a next step, we probably need to check this from a different machine/user and see how the change affects reproducible builds.

**Comment by mariaschett** [comment]
So using rootless docker worked great for hello-transparent-release. 

I think we can close this discussion and not rely on #123 .
